src/librustc_resolve/late/diagnostics.rs

```diff
@@ -558,15 +569,74 @@ impl<'a> LateResolutionVisitor<'a, '_, '_> {
                             ),
                         );
                     }
-                    suggested = true;
                 }
-                _ => {}
-            }
-            if !suggested {
-                if let Some(span) = self.r.opt_span(def_id) {
-                    err.span_label(span, &format!("`{}` defined here", path_str));
+                PathSource::TupleStruct(span) => {
+                    // err.replace_span_with(span);
+                    if let Some(span) = self.r.opt_span(def_id) {
+                        err.span_label(span, &format!("`{}` defined here", path_str));
+                    }
+                    let (fields, appl) = match self.r.field_names.get(&def_id) {
+                        Some(fields) => (
+                            fields
+                                .iter()
+                                .map(|f| f.node.to_string())
+                                .collect::<Vec<String>>()
+                                .join(", "),
+                            Applicability::MachineApplicable,
+                        ),
+                        None => ("/* fields */".to_string(), Applicability::HasPlaceholders),
+                    };
+                    err.span_suggestion(
+                        span,
+                        "use struct pattern syntax instead",
+                        format!("{} {{{pad}{}{pad}}}", path_str, fields, pad = pad),
+                        appl,
+                    );
                 }
-                err.span_label(span, format!("did you mean `{} {{ /* fields */ }}`?", path_str));
+                PathSource::Expr(
+                    _,
+                    Some(Expr { span: call_span, kind: ExprKind::Call(expr, _), .. }),
+                ) if expr.span == span => {
+                    if let Some(span) = self.r.opt_span(def_id) {
+                        err.span_label(span, &format!("`{}` defined here", path_str));
+                    }
+                    let fields = match self.r.field_names.get(&def_id) {
+                        Some(fields) => fields
+                            .iter()
+                            .map(|f| format!("{}: val", f.node))
+                            .collect::<Vec<String>>()
+                            .join(", "),
+                        None => "/* fields */".to_string(),
+                    };
+                    err.span_suggestion(
+                        *call_span,
+                        "use struct literal syntax instead",
+                        format!("{} {{{pad}{}{pad}}}", path_str, fields, pad = pad),
+                        Applicability::HasPlaceholders,
+                    );
+                }
+                PathSource::Expr(Some(Expr { kind: ExprKind::Path(..), .. }), _)
```

Comment:
I don't understand the point of the `ExprKind::Path` condition, it's always true.
`PathSource::Expr` is always constructed with `ExprKind::Path` in this PR (except for the `PathSource::Expr(None, None)`, which is a bug).